In [1]:
import os
from pyspark.sql import SparkSession

In [2]:
import matplotlib.pyplot as plt
import numpy as np
from pyspark.sql.functions import when, col, sum

Matplotlib created a temporary cache directory at /scratch/abhattacharya6/job_49205195/matplotlib-5odd297x because the default path (/home/jovyan/.cache/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.


In [3]:
# Template: Adjust TOTAL_MEMORY and TOTAL_CORES to match your allocation
TOTAL_MEMORY = 128  # GB - from your Jupyter request
TOTAL_CORES = 64     # from your Jupyter request
DRIVER_MEMORY = 2   # GB - fixed, small (driver doesn't process data)

num_executors = TOTAL_CORES - 1
executor_memory = (TOTAL_MEMORY - DRIVER_MEMORY) // num_executors

spark = SparkSession.builder \
    .config("spark.driver.memory", f"{DRIVER_MEMORY}g") \
    .config("spark.executor.memory", f"{executor_memory}g") \
    .config("spark.executor.instances", num_executors) \
    .config("spark.sql.shuffle.partitions", "200") \
    .config("spark.memory.offHeap.enabled", "true") \
    .config("spark.memory.offHeap.size", "4g") \
    .config("spark.sql.parquet.columnarReaderBatchSize", "1024") \
    .getOrCreate()

In [ ]:
#os.environ["HF_TOKEN"] = "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"

In [ ]:
#This cell is used to download the dataset to scratch
# import os
# from huggingface_hub import snapshot_download

# base = os.path.expanduser("~/scratch")

# os.environ["TMPDIR"] = f"{base}/tmp"
# os.environ["TEMP"] = f"{base}/tmp"
# os.environ["TMP"] = f"{base}/tmp"

# os.makedirs(f"{base}/tmp", exist_ok=True)

# hf_token = "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"

# pth = snapshot_download(
#     repo_id="mercari-us/merrec",
#     repo_type="dataset",
#     token=hf_token,
#     cache_dir=f"{base}/hf_cache",
#     local_dir=f"{base}/merrec_snapshot",
#     local_dir_use_symlinks=False
# )

# print("Downloaded to:", pth)

In [ ]:
# df = spark.read.option(
#     "recursiveFileLookup", "true"
# ).option(
#     "pathGlobFilter", "*.parquet"
# ).parquet(f"{base}/merrec_snapshot")

# df.printSchema()
# df.show(5)

In [4]:
base = "/home/abhattacharya6/scratch/merrec_snapshot"

df = spark.read.parquet(f"{base}/20230901/000000000000.parquet")

tiny_df = df.select("item_id", "price").limit(1000)

tiny_df.show(5)

+---------+-----+
|  item_id|price|
+---------+-----+
|234718099| 30.0|
| 80353410| 28.0|
|101459145| 60.0|
|141008967| 18.0|
|141008967| 18.0|
+---------+-----+
only showing top 5 rows



In [5]:
tiny_df.createOrReplaceTempView("sampled")

exp_df_1 = spark.sql("""
    SELECT item_id, MIN(price) AS Price
    FROM sampled
    GROUP BY item_id
""")

exp_df_1.show(5)

+---------+-----+
|  item_id|Price|
+---------+-----+
|234718099| 30.0|
| 80353410| 28.0|
|101459145| 60.0|
|141008967| 18.0|
| 49192775|14.99|
+---------+-----+
only showing top 5 rows



In [ ]:
# Use this for loading the dataset in

# base = "/home/abhattacharya6/scratch"

# df = spark.read.option(
#     "recursiveFileLookup", "true"
# ).option(
#     "pathGlobFilter", "*.parquet"
# ).parquet(f"{base}/merrec_snapshot")

In [ ]:
df.coalesce(64)

In [ ]:
df.printSchema()

In [ ]:
df.createOrReplaceTempView("merrec")

In [ ]:
df.limit(5).show()

# 3a

In [ ]:
df.select("price").count()

# 3c

In [ ]:
#df.select([
#    sum(when(col(c).isNull(),1).otherwise(0)).alias(c)
#    for c in df.columns
#]).show()
#See README for output (Memory Issues)

# 4

In [13]:
sample_df = df.select("item_id", "price").limit(50000)

sample_df.createOrReplaceTempView("sampled")

exp_df_1 = spark.sql("""
    SELECT item_id, MIN(price) AS Price
    FROM sampled
    GROUP BY item_id
""")

exp_df_1.show(5)

AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `item_id` cannot be resolved. Did you mean one of the following? [`event_id`, `label`, `price`, `c0_name`, `c1_name`].;
'Project ['item_id, price#1875]
+- Project [price#1875, event_id#1811, c0_name#1867, c1_name#1868, item_condition_name#1869, price_clipped#1881, log_price#1888, sqrt_price#1896, CASE WHEN (event_id#1811 = buy_comp) THEN 1.0 ELSE 0.0 END AS label#1905]
   +- Project [price#1875, event_id#1811, c0_name#1867, c1_name#1868, item_condition_name#1869, price_clipped#1881, log_price#1888, SQRT(price_clipped#1881) AS sqrt_price#1896]
      +- Project [price#1875, event_id#1811, c0_name#1867, c1_name#1868, item_condition_name#1869, price_clipped#1881, LOG1P(price_clipped#1881) AS log_price#1888]
         +- Project [price#1875, event_id#1811, c0_name#1867, c1_name#1868, item_condition_name#1869, CASE WHEN (price#1875 < cast(0 as double)) THEN 0.0 WHEN (price#1875 > cast(5000 as double)) THEN 5000.0 ELSE price#1875 END AS price_clipped#1881]
            +- Project [cast(price#1815 as double) AS price#1875, event_id#1811, c0_name#1867, c1_name#1868, item_condition_name#1869]
               +- Project [price#1815, event_id#1811, coalesce(c0_name#1816, cast(Unknown as string)) AS c0_name#1867, coalesce(c1_name#1818, cast(Unknown as string)) AS c1_name#1868, coalesce(item_condition_name#1825, cast(Unknown as string)) AS item_condition_name#1869]
                  +- GlobalLimit 500000
                     +- LocalLimit 500000
                        +- Sample 0.0, 5.0E-4, false, 42
                           +- Project [price#1815, event_id#1811, c0_name#1816, c1_name#1818, item_condition_name#1825]
                              +- Relation [user_id#1806L,stime#1807,session_id#1808,sequence_id#1809,sequence_length#1810L,event_id#1811,item_id#1812L,product_id#1813,name#1814,price#1815,c0_name#1816,c0_id#1817L,c1_name#1818,c1_id#1819L,c2_name#1820,c2_id#1821L,brand_name#1822,brand_id#1823L,item_condition_id#1824L,item_condition_name#1825,size_name#1826,size_id#1827L,color#1828,shipper_id#1829L,shipper_name#1830] parquet


In [ ]:
sample_df = spark.sql("SELECT item_id, price FROM merrec TABLESAMPLE (1 PERCENT)")
sample_df.createOrReplaceTempView("sampled")
exp_df_1 = spark.sql("SELECT item_id, MIN(price) as Price FROM sampled GROUP BY item_id")
exp_df_1.limit(5).show()

In [ ]:
bins,counts = exp_df_1.select("price").rdd.flatMap(lambda x: x).histogram(50)
plt.bar(bins[:-1], counts, width=bins[1]-bins[0])
plt.xlabel("price")
plt.ylabel("count")
plt.title("Price histogram")
plt.show()

In [ ]:
exp_df_2 = spark.sql("SELECT event_id, count(event_id) AS count FROM merrec GROUP BY event_id ORDER BY count DESC")
exp_df_2.show()

In [ ]:
data = exp_df_2.select("event_id", "count").collect()

x = [row["event_id"] for row in data]
y = [row["count"] for row in data]

plt.bar(x,y, width=0.8)
plt.xlabel("event_id")
plt.ylabel("count")
plt.xticks(rotation='vertical')
plt.yticks(np.arange(0,1054294103,100000000))
plt.ticklabel_format(style='plain', axis='y', useOffset=False)
plt.title('event_id Class Counts')
plt.show()

In [ ]:
exp_df_3 = spark.sql("SELECT c0_name, count(c0_name) AS count FROM merrec GROUP BY c0_name ORDER BY count DESC")
exp_df_3.show()

In [ ]:
data = exp_df_3.select("c0_name", "count").collect()

x = [row["c0_name"] for row in data]
y = [row["count"] for row in data]

plt.bar(x,y, width=0.8)
plt.xlabel("c0_name")
plt.ylabel("count")
plt.xticks(rotation='vertical')
plt.ticklabel_format(style='plain', axis='y', useOffset=False)
plt.title('c0_name Class Counts')
plt.show()

In [ ]:
sample_df_2 = spark.sql("SELECT event_id, price FROM merrec WHERE event_id ='item_like' or event_id = 'item_view'")
sample_df_2 = sample_df_2.withColumn("label",when(col("event_id") == "item_like", 1).otherwise(0))
sample_df_2.createOrReplaceTempView("sampled2")
exp_df_4 = spark.sql("SELECT FLOOR(price / 10) * 10 AS price_bin, AVG(label) AS avg_label FROM sampled2 GROUP BY FLOOR(price / 10) * 10 ORDER BY price_bin")

In [ ]:
exp_df_4.limit(5).show()

In [ ]:
data = exp_df_4.select("price_bin", "avg_label").collect()

x = [row["price_bin"] for row in data]
y = [row["avg_label"] for row in data]

plt.plot(x,y)
plt.xlabel("price bin")
plt.ylabel("Like Percentage")
plt.title('Average Rate of Likes Among Like and View Events per price')
plt.show()

In [ ]:
import requests
import pandas as pd

# Get the active Spark Context and URL
sc = spark.sparkContext
url = f"{sc.uiWebUrl}/api/v1/applications/{sc.applicationId}/executors"

# Fetch the executor data from the API
response = requests.get(url)
executors = response.json()

# Format into a readable DataFrame
df = pd.DataFrame(executors)[['id', 'totalCores', 'maxMemory', 'activeTasks', 'isActive']]
df['maxMemory_GB'] = (df['maxMemory'] / (1024**3)).round(2)
df

### Preprocessing Summary

The preprocessing pipeline was implemented using Spark MLlib transformers and DataFrame operations.
Preprocessing process:

1. Selected a subset of relevant columns from the original parquet dataset.
2. Sampled the dataset to reduce computation time and avoid memory issues
3. Filled missing categorical values using placeholder strings.
4. Used Imputer to handle missing numeric values in price.
5. Created additional engineered features:
   - log_price
   - sqrt_price
6. Encoded categorical columns using:
   - StringIndexer
   - OneHotEncoder
7. Combined all features using VectorAssembler.
8. Applied MinMaxScaler to normalize the final feature vector.
9. Split data into training, validation, and test sets.


In [9]:
#Milestone 3 - Preprocessing
from pyspark.ml import Pipeline
from pyspark.ml.feature import (
    StringIndexer, OneHotEncoder, VectorAssembler,
    MinMaxScaler, Imputer
)
from pyspark.sql import functions as F
from pyspark.sql.types import DoubleType

cols = [
    "price", "event_id", "c0_name", "c1_name",
    "item_condition_name"
]

df = spark.read \
    .option("recursiveFileLookup", "true") \
    .option("pathGlobFilter", "*.parquet") \
    .parquet("/home/abhattacharya6/scratch/merrec_snapshot") \
    .select(*cols)

#sample to avoid memory issues
df = df.sample(False, 0.0005, seed=42)
df = df.limit(500000)

#clean data
df = df.fillna({
    "c0_name": "Unknown",
    "c1_name": "Unknown",
    "item_condition_name": "Unknown"
})

df = df.withColumn("price", F.col("price").cast(DoubleType()))

#feature engineering
df = df.withColumn(
    "price_clipped",
    F.when(F.col("price") < 0, 0.0)
     .when(F.col("price") > 5000, 5000.0)
     .otherwise(F.col("price"))
)

df = df.withColumn("log_price", F.log1p("price_clipped"))
df = df.withColumn("sqrt_price", F.sqrt("price_clipped"))

#binary label
df = df.withColumn(
    "label",
    F.when(F.col("event_id") == "buy_comp", 1.0).otherwise(0.0)
)

#encoding
cat_cols = ["c0_name", "c1_name", "item_condition_name"]

indexers = [
    StringIndexer(
        inputCol=c,
        outputCol=f"{c}_idx",
        handleInvalid="keep"
    ) for c in cat_cols
]

encoder = OneHotEncoder(
    inputCols=[f"{c}_idx" for c in cat_cols],
    outputCols=[f"{c}_ohe" for c in cat_cols]
)

#imputer
imputer = Imputer(
    inputCols=["price_clipped"],
    outputCols=["price_imputed"]
)

#assembler
numeric_cols = [
    "price_imputed",
    "log_price",
    "sqrt_price"
]

assembler = VectorAssembler(
    inputCols=numeric_cols + [f"{c}_ohe" for c in cat_cols],
    outputCol="features_raw",
    handleInvalid="skip"
)

#scaler
scaler = MinMaxScaler(
    inputCol="features_raw",
    outputCol="features"
)

#pipeline
pipeline = Pipeline(stages=
    indexers + [encoder, imputer, assembler, scaler]
)

train_df, val_df, test_df = df.randomSplit([0.7, 0.15, 0.15], seed=42)

model = pipeline.fit(train_df)

train_final = model.transform(train_df)
val_final = model.transform(val_df)
test_final = model.transform(test_df)

#show preprocessed data
train_final.select(
    "event_id",
    "c0_name",
    "c1_name",
    "item_condition_name",
    "features",
    "label"
).show(10, truncate=False)

+--------------------+-------------+----------------------+-------------------+--------------------------------+-----+
|event_id            |c0_name      |c1_name               |item_condition_name|features                        |label|
+--------------------+-------------+----------------------+-------------------+--------------------------------+-----+
|buy_start           |Women        |Tops & blouses        |Good               |(290,[3,25,284],[1.0,1.0,1.0])  |0.0  |
|item_add_to_cart_tap|Arts & Crafts|Fiber art             |Like new           |(290,[16,217,286],[1.0,1.0,1.0])|0.0  |
|item_add_to_cart_tap|Electronics  |Media                 |Good               |(290,[8,45,284],[1.0,1.0,1.0])  |0.0  |
|item_add_to_cart_tap|Electronics  |Video games & consoles|Good               |(290,[8,30,284],[1.0,1.0,1.0])  |0.0  |
|item_add_to_cart_tap|Home         |Kitchen Drinkware     |Like new           |(290,[7,59,286],[1.0,1.0,1.0])  |0.0  |
|item_add_to_cart_tap|Home         |Seasonal dec

## Fitting Analysis

In [10]:
# Training first model

from pyspark.ml.classification import DecisionTreeClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

dt1 = DecisionTreeClassifier(
    labelCol="label",
    featuresCol="features",
    maxDepth=5
)

dt_model1 = dt1.fit(train_final)

pred1_train = dt_model1.transform(train_final)
pred1_test = dt_model1.transform(test_final)

evaluator = MulticlassClassificationEvaluator(
    labelCol="label",
    predictionCol="prediction",
    metricName="accuracy"
)

train_acc_1 = evaluator.evaluate(pred1_train)
test_acc_1 = evaluator.evaluate(pred1_test)

print("Model 1 Train Accuracy:", train_acc_1)
print("Model 1 Test Accuracy:", test_acc_1)

Model 1 Train Accuracy: 0.9991396681577179
Model 1 Test Accuracy: 0.9990017037589181


In [11]:
# Second hyperparamter model

dt2 = DecisionTreeClassifier(
    labelCol="label",
    featuresCol="features",
    maxDepth=10
)

dt_model2 = dt2.fit(train_final)

pred2_train = dt_model2.transform(train_final)
pred2_test = dt_model2.transform(test_final)

train_acc_2 = evaluator.evaluate(pred2_train)
test_acc_2 = evaluator.evaluate(pred2_test)

print("Model 2 Train Accuracy:", train_acc_2)
print("Model 2 Test Accuracy:", test_acc_2)

Model 2 Train Accuracy: 0.9991425264030411
Model 2 Test Accuracy: 0.9989883931423703


In [12]:
# Comparison table

comparison = spark.createDataFrame([
    ("Decision Tree", "maxDepth=5", train_acc_1, test_acc_1),
    ("Decision Tree", "maxDepth=10", train_acc_2, test_acc_2)
], ["Model", "Hyperparameters", "Train Accuracy", "Test Accuracy"])

comparison.show()

+-------------+---------------+------------------+------------------+
|        Model|Hyperparameters|    Train Accuracy|     Test Accuracy|
+-------------+---------------+------------------+------------------+
|Decision Tree|     maxDepth=5|0.9991396681577179|0.9990017037589181|
|Decision Tree|    maxDepth=10|0.9991425264030411|0.9989883931423703|
+-------------+---------------+------------------+------------------+



The Decision Tree models achieved very high training and testing accuracy. The maxDepth=5 model achieved approximately 99.91% training accuracy and 99.90% testing accuracy, while the maxDepth=10 model achieved similar results with slightly lower testing accuracy.

The small gap between training and testing accuracy suggests that the models generalize well and are not significantly overfitting. Although the deeper tree performed slightly better on the training data, the shallower tree generalized slightly better on unseen data.

Overall, the maxDepth=5 model performed best because it achieved nearly identical performance with a simpler model structure. For Milestone 4, ensemble models such as Random Forests and Gradient Boosted Trees could be explored to further improve robustness and predictive performance.